### PR TITLE
facade crate

### PR DESCRIPTION
**Pull request by insipx**
converging the two codepaths into one crate (name subject to change)

### CODE REVIEW

File: facade/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+[package]
+name = "facade"
```
[the remark this answers is not in the document]
Comment (insipx) [reply facade/Cargo.toml]: renamed to 'desub'

File: desub/src/lib.rs [hunk 2]
```diff
@@ -82,9 +83,13 @@ impl<T: TypeDetective> Decoder<T> {
 	}
 }
 
+/// Decoder which does not resolve any types of metadata v13 and below.
+/// To be used with newer chains where legacy types are not required.
 pub struct InfoDecoder(Decoder<NoLegacyTypes>);
-impl Default for InfoDecoder {
-	fn default() -> Self {
+impl InfoDecoder {
+    // No sensible default b/c of Chain::Custom("none")
```
Comment (dvdplm) [inline desub/src/lib.rs:90]: funny whitespace here?

**Review by dvdplm — APPROVED** [review]

File: core/Cargo.toml [hunk 3]
```diff
@@ -17,9 +17,9 @@ dyn-clone = "1.0"
 hex = "0.4"
 bitvec = { version = "0.20.2", features = ["serde", "alloc"] }
 
-pallet-democracy = { package = "pallet-democracy", git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.9" }
-primitives = { package = "sp-core",  git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.9" }
-runtime-primitives = { package = "sp-runtime", git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.9" }
+pallet-democracy = { package = "pallet-democracy", git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.12" }
+primitives = { package = "sp-core",  git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.12" }
+runtime-primitives = { package = "sp-runtime", git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.12" }
 frame-metadata = { package = "frame-metadata", git = "https://github.com/paritytech/frame-metadata", branch = "main", features = ["legacy"] }
```
Comment (dvdplm) [inline core/Cargo.toml:23]: Do you know what is blocking `frame-metadata` from publication?

Comment (insipx) [reply core/Cargo.toml:23]: have no idea; i just pinged andrew about it though

File: desub/src/lib.rs (hunk 2 above)
Comment (insipx) [reply desub/src/lib.rs]: my rust version was pre-2021 and i didn't even realize the linter couldn't run :(

all fixed

File: core/Cargo.toml (hunk 3 above)
Comment (ascjones) [reply core/Cargo.toml:23]: Fine by me for you to go ahead and create and push a `14.1` release. Substrate won't need to be updated since it's just the addition of the legacy features.

File: desub/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+name = "desub"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+
+desub_legacy = { path = "../core/", package = "desub-core" }
+core_v14 = { path = "../core_v14/", package = "core_v14" }
```
Comment (jsdw) [inline desub/Cargo.toml:11]: I wonder whether we should consistify these names to, maybe:

```
core_legacy
core_v14
```

or something (just so it's obvious that they are closely related when using).

File: desub/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+			self.legacy_decoder.register_version(version, LegacyDesubMetadata::from_runtime_metadata(metadata.1)?)?;
+		}
+		Ok(())
+	}
+}
+
+/// Decoder which does not resolve any types of metadata v13 and below.
+/// To be used with newer chains where legacy types are not required.
+pub struct InfoDecoder(Decoder<NoLegacyTypes>);
+impl InfoDecoder {
```
Comment (jsdw) [inline desub/src/lib.rs:89]: Supernit: there's a space between struct and impl below but not here :)

File: desub/src/error.rs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+// You should have received a copy of the GNU General Public License
+// along with substrate-desub.  If not, see <http://www.gnu.org/licenses/>.
+
+use core_v14::{decoder::DecodeError, metadata::MetadataError};
+use desub_legacy::{decoder::metadata::Error as LegacyMetadataError, Error as LegacyError};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum Error {
+	#[error(transparent)]
```
Comment (jsdw) [inline desub/src/error.rs:23]: I hadn't noticed the `transparent` option before; mental note made :)

File: desub/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+		let decoder = Decoder::new(NoLegacyTypes, Chain::Custom("none".to_string()));
+		Self(decoder)
+	}
+}
+
+/// Decoder that includes all Polkadot JS type definitions in addition to
+/// V14 Metadata decoding.
+/// This is useful if historical type network data that existed pre-v14 is required.
+#[cfg(feature = "polkadot-js")]
+pub struct PolkadotJsDecoder(Decoder<PolkadotJsResolver>);
```
Comment (jsdw) [inline desub/src/lib.rs:102]: What do I do with these decoder structs once I've created them?

File: desub/Cargo.toml (hunk 4 above)
Comment (jsdw) [reply desub/Cargo.toml:11]: now that we've agreed on names, `desub-current` and `desub-legacy` I geuss would be the most future proof choices :)

**Comment by insipx** [comment]
I refactored to remove the separate structs and instead rely more directly on the feature flags. I went through a bit of an identity crisis in the middle of designing this crate since it's mostly created from my experience/use case for Archive, and maybe won't be as directed for Signer/projects that want less of a stateful approach.

The biggest issue I see is that if you just want to use V14 (no feature flags, default `new()` method) you will still be compiling with all the `legacy` crate code. If code size is not an issue this doesn't matter very much, but if we ever want to for instance have an initiative for compiling desub to wasm or something that would have to be re-thought. 


This PR actually turned out to be a bit larger than I anticipated. A few things I changed since you looked guys looked at it:

- Added `Serialize`, `Deserialize` methods on all V14 generic structs (Extrinsic, ExtrinsicSignature etc.).
- Added `desub-common` crate where we can stuff types that are used in both current and legacy code. currently this houses the remote implementation of `MultiAddress` as well as a `SpecVersion` type alias.
- Used the new `facade` crate in tx-decoder

File: desub/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+		}
+		Ok(())
+	}
+
+	pub fn decode_extrinsics(&self, version: SpecVersion, mut data: &[u8]) -> Result<Value, Error> {
+		if self.current_decoder.contains_key(&version) {
+			let decoder = self.current_decoder.get(&version).expect("Checked if key is contained; qed");
+			match decoder.decode_extrinsics(&mut data) {
+				Ok(v) => Ok(serde_json::to_value(&v)?),
+				Err((ext, e)) => Err(Error::V14{ source: e, ext: ext.into_iter().map(Extrinsic::into_owned).collect() }),
```
[the remark this answers is not in the document]
Comment (jsdw) [reply desub/src/lib.rs]: This is just because we might hit an error after some amount of extrinsics, and so I wanted to return all of the extrinsics parsed so far. If you like, you can just discard the extrinsics and report the error for the block, or we know which extrinsic failed by counting the successful ones?

I dont' think it makes sense to embed them into the error, because it wouldn't make sense for most errors.

I did try having the `decode_extrinsics` function return an iterator (which would allow you to collect extrinsics and stop when you hit an error (which would leave you with the same result), but it would have added a bit much complexity so I scrapped it!

File: bin/tx-decoder/src/app.rs [hunk 9]
```diff
@@ -128,8 +126,8 @@ impl<'a> AppState<'a> {
 	fn decode(decoder: &Decoder, block: BlockModel, spec: SpecVersion, errors: &mut Vec<String>) -> Result<(), Error> {
 		log::debug!("Decoding block {}, spec_version {}, ext length {}", block.block_num, spec, block.ext.len());
 		match decoder.decode_extrinsics(spec, &block.ext) {
-			Err(e) => {
-				let e = e.context(format!("Failed to decode block {}", block.block_num));
+			e @ Err(_) => {
+				let e = e.context(format!("Failed to decode block {}", block.block_num)).unwrap_err();
```
Comment (jsdw) [inline bin/tx-decoder/src/app.rs:130]: I'm curious to know why this change was necessary? does the `context` method on the trait give you something that the `context` method on `anyhow::Error` doesn't?

File: core_v14/src/value/mod.rs [hunk 10]
```diff
@@ -34,7 +34,7 @@ use std::fmt::Debug;
 /// sequence, array and composite types can all be represented with [`Composite`]. Only enough information
 /// is preserved here to construct a valid value for any type that we know about, and be able to verify
 /// that a given value is compatible with some type (see the [`scale_info`] crate), if we have both.
-#[derive(Clone, PartialEq)]
```
Comment (jsdw) [inline core_v14/src/value/mod.rs:37]: I guess these are serialized to be stored in the database?

My only hesitation here is that the serialize and deserialize impls aren't symmetric. I wonder whether it would be better to write serialize impls that transform like so:

`Value::Composite(Composite::Unnamed(vec![values]))` => sequence of values.
`Value::Composite(Composite::Named(vec![keyvalues]))` => map of key+value pairs.
`Value::Variant(..)` => serde data model variant.
`Value::Primitive(Primitive::Foo(..)) => serialize straight to the primitive value in question.
`Value::BitSequence(..)` => serialize straight to the bitsequence?

This would mirror Deserialize more closely, which attempt to deserialize based into the relevant structure based on the type of Value (whereas by default it would always try to deserialize into an enum variant containing eg a composite which is another enum variant and so on).

What do you think?

Comment (jsdw) [reply core_v14/src/value/mod.rs:37]: If you like, I can work on said impls this afternoon!

Comment (jsdw) [reply core_v14/src/value/mod.rs:37]: PR here: https://github.com/paritytech/desub/pull/69

File: bin/tx-decoder/src/app.rs (hunk 9 above)
Comment (insipx) [reply bin/tx-decoder/src/app.rs:130]: since i switched tx-decoder to use the facade crate, the method returns a thiserror::Error which doesn't have a direct `context` method like `anyhow::Error` does. Instead the context method is accessed through the `Context` trait on the general `Result` type

Comment (jsdw) [reply bin/tx-decoder/src/app.rs:130]: Aah ok, so you're using the trait to turn it into an `anyhow::Error` with context!

I did have a look and found that you could perhaps do something like:

`anyhow::Error::new(e).context(format!("Failed to decode block {}", block.block_num))` to avoid the result unwrapping thing if you like.

**Review by jsdw — APPROVED** [review]
